packages/grid/x-data-grid/src/internals/hooks/features/columns/gridColumnsUtils.ts

```diff
@@ -31,6 +31,130 @@ export const computeColumnTypes = (customColumnTypes: GridColumnTypesRecord = {}
   return mergedColumnTypes;
 };
 
+/**
+ * Computes width for flex columns.
+ * Based on CSS Flexbox specification:
+ * https://drafts.csswg.org/css-flexbox-1/#resolve-flexible-lengths
+ */
+export function computeFlexColumnsWidth({
+  initialFreeSpace,
+  totalFlexUnits,
+  flexColumns,
+}: {
+  initialFreeSpace: number;
+  totalFlexUnits: number;
+  flexColumns: {
+    field: GridColDef['field'];
+    flex?: number;
+    minWidth?: number;
+    maxWidth?: number;
+  }[];
+}) {
+  const flexColumnsLookup: {
+    all: Record<
+      GridColDef['field'],
+      {
+        flex: number;
+        computedWidth: number;
+        frozen: boolean;
+      }
+    >;
+    frozenFields: GridColDef['field'][];
+    freeze: (field: GridColDef['field']) => void;
+  } = {
+    all: {},
+    frozenFields: [],
+    freeze: function freeze(field: GridColDef['field']) {
+      const value = this.all[field];
+      if (value && value.frozen !== true) {
+        this.all[field].frozen = true;
+        this.frozenFields.push(field);
+      }
+    },
+  };
+
+  // Step 5 of https://drafts.csswg.org/css-flexbox-1/#resolve-flexible-lengths
+  function loopOverFlexItems() {
+    // 5a: If all the flex items on the line are frozen, free space has been distributed.
+    if (flexColumnsLookup.frozenFields.length === flexColumns.length) {
+      return;
+    }
+
+    const violationsLookup: {
+      min: Record<GridColDef['field'], boolean>;
+      max: Record<GridColDef['field'], boolean>;
+    } = { min: {}, max: {} };
+
+    let remainingFreeSpace = initialFreeSpace;
+    let flexUnits = totalFlexUnits;
+    let totalViolation = 0;
+
+    // 5b: Calculate the remaining free space
+    flexColumnsLookup.frozenFields.forEach((field) => {
+      remainingFreeSpace -= flexColumnsLookup.all[field].computedWidth;
+      flexUnits -= flexColumnsLookup.all[field].flex!;
+    });
+    for (let i = 0; i < flexColumns.length; i += 1) {
+      const column = flexColumns[i];
+
+      if (
+        flexColumnsLookup.all[column.field] &&
```

Comment:
Off-topic, but I think we should set `noImplicitAny: true` as soon as possible.
At some point I had a runtime error here, because I've used `flexColumnsLookup[column.field]` instead of `flexColumnsLookup.all[column.field]` and TypeScript didn't pick it up :/

There's an opened issue for it https://github.com/mui/mui-x/issues/479
@joserodolfofreitas can we add it to our backlog?

Reply:
I worked on it last summer
If you have time you can resume the effort, it's a pain :laughing:

Reply:
> There's an opened issue for it #479 @joserodolfofreitas can we add it to our backlog?
Done!
Good find, @cherniavskii!

Reply:
@cherniavskii #2505 for instance
I enable the rules, fix some TS errors, disable the rule and open a PR.
Doing everything in a single PR is too long.

Reply:
Yeah, I was thinking about doing it gradually as well.
Starting with one package (I'm thinking `x-data-grid-pro`) and eventually reaching the point where we can enable the flag for this specific package (since each package has its own `tsconfig.json`).

Reply:
Yes now that we have distinct `tsconfig.json` per package it should be easier to advance on that topic